pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
 	c.queue.Add(podKey)
 }
 
-func (c *Controller) Worker() {
+func (c *NPLController) checkDeletedSvc(obj interface{}) (*corev1.Service, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+	}
+	svc, ok := deletedState.Obj.(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Service: %v", deletedState.Obj)
+	}
+	return svc, nil
+}
+
+func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both sets of Pods (corresponding to old and new selector) need to be considered
+
+	// Donot push to queue if 1) Service ResourceVersions don't change OR
+	// 2) Service spec selectors AND 3) NPLEnabledAnnotationKey values don't change
+	newSvc := newObj.(*corev1.Service)
+	oldSvc := oldObj.(*corev1.Service)
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion ||
+		reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) &&
+			oldSvc.Annotations[NPLEnabledAnnotationKey] == newSvc.Annotations[NPLEnabledAnnotationKey] {
+		return
+	}
+
+	// disjunctive union of Pods from both Service sets
+	oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
```

Comment:
As an optimization, can we get both old and new sets only when selectors changed?

Like:
```
if oldSvc.ResourceVersion == newSvc.ResourceVersion {
    return
}
selectorsChanged := !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector)
if !selectorsChanged && oldSvc.Annotations[NPLEnabledAnnotationKey] == newSvc.Annotations[NPLEnabledAnnotationKey] {
    return
}
newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
if selectorsChanged {
    ...
}
```

Reply:
This is covered in the conditional (the negation of it), if the selectors are same we return.

Reply:
I think if the selectors do not change, but NPL annotation changes, we will still get here? I meant in this case no need to compute both old and new Pod sets, since they are equal.

Reply:
Actually, probably we better to handle annotation change and selector change separately. If annotation changes, we need not to check selectors at all, but should just either remove old Pod sets or add new Pod sets.

Reply:
>no need to compute both old and new Pod sets, since they are equal.

So if the annotation changes from valid (oldSvc) to invalid (newSvc), then the newSvc will map to no pods since it does not have the valid NPL annotation in it (see getPodsFromService). So we'll end up queueing only the pods corresponding to the service with the valid annotation, i.e. oldSvc.

>probably we better to handle annotation change and selector change separately

I didn't quite get the second comment. The code snippet provided by you is pretty much what I have implemented, just squashed into a single conditional block. Are you suggesting to break down the conditional written above into separate blocks, with each block returning if no change is detected? Something like this:

```
if oldSvc.ResourceVersion == newSvc.ResourceVersion {
    return
}

if oldSvc.Annotations[NPLEnabledAnnotationKey] != newSvc.Annotations[NPLEnabledAnnotationKey] {
    newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
    podKeys := oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
    for podKey := range podKeys {
        c.queue.Add(podKey)
    }
    return
}

if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
    newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
    podKeys := oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
    for podKey := range podKeys {
        c.queue.Add(podKey)
    }
    return
}
```

Reply:
Ok, I got what you meant now. Then if we want to optimize (like the way in the above code block), we can just save the annotation lookup in getPodsFromService(), creating empty Pod sets, Set.Difference() and Set.Union(). I can agree these are not of big overhead.
I always like more (habit from my datapath experience), but I understand you might like to keep code simpler. So, I would let you decide.